spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/metadata/JdbcMetadataStore.java

```diff
@@ -0,0 +1,198 @@
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Jdbc-based implementation of {@link MetadataStore}.
+ *
+ * @author Bojan Vukasovic
+ * @since 5.0
+ */
+public class JdbcMetadataStore implements ConcurrentMetadataStore {
+
+	/**
+	 * Default value for the table prefix property.
+	 */
+	public static final String DEFAULT_TABLE_PREFIX = "INT_";
+
+	private final JdbcOperations jdbcTemplate;
+
+	private volatile Map<Query, String> queryCache = new HashMap<>();
+
+	private enum Query {
+		GET_VALUE("SELECT METADATA_VALUE FROM %PREFIX%METADATA_STORE WHERE METADATA_KEY=?"),
+		GET_VALUE_FOR_UPDATE("SELECT METADATA_VALUE FROM %PREFIX%METADATA_STORE WHERE METADATA_KEY=? FOR UPDATE"),
+		REPLACE_VALUE("UPDATE %PREFIX%METADATA_STORE SET METADATA_VALUE=? WHERE METADATA_KEY=? AND METADATA_VALUE=?"),
+		REPLACE_VALUE_BY_KEY("UPDATE %PREFIX%METADATA_STORE SET METADATA_VALUE=? WHERE METADATA_KEY=?"),
+		REMOVE_VALUE("DELETE FROM %PREFIX%METADATA_STORE WHERE METADATA_KEY=?"),
+		PUT_IF_ABSENT_VALUE("INSERT INTO INT_METADATA_STORE(METADATA_KEY, METADATA_VALUE) "
+				+ "SELECT ?, ? FROM INT_METADATA_STORE WHERE METADATA_KEY=? HAVING COUNT(*)=0");
+
+		private String sql;
+
+		Query(String sql) {
+			this.sql = sql;
+		}
+
+		public String getSql() {
+			return this.sql;
+		}
+	}
+
+	/**
+	 * Instantiate a {@link JdbcMetadataStore} using provided dataSource {@link DataSource}.
+	 * @param dataSource a {@link DataSource}
+	 */
+	public JdbcMetadataStore(DataSource dataSource) {
+		this(new JdbcTemplate(dataSource));
+	}
+
+	/**
+	 * Instantiate a {@link JdbcMetadataStore} using provided jdbcOperations {@link JdbcOperations}.
+	 * @param jdbcOperations a {@link JdbcOperations}
+	 */
+	public JdbcMetadataStore(JdbcOperations jdbcOperations) {
+		Assert.notNull(jdbcOperations, "'dataSource' must not be null");
+		this.jdbcTemplate = jdbcOperations;
+	}
+
+	@Override
+	@Transactional(propagation = Propagation.REQUIRES_NEW)
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' cannot be null");
+		Assert.notNull(value, "'value' cannot be null");
+		while (true) {
+			//try to insert if does not exists
+			int affectedRows = this.jdbcTemplate.update(getQuery(Query.PUT_IF_ABSENT_VALUE), ps -> {
+				ps.setString(1, key);
+				ps.setString(2, value);
+				ps.setString(3, key);
+			});
+			if (affectedRows > 0) {
+				//it was not in the table, so we just inserted it
+				return null;
+			}
+			else {
+				//value should be in table. try to return it
+				try {
+					return this.jdbcTemplate.queryForObject(getQuery(Query.GET_VALUE), String.class, key);
+				}
+				catch (EmptyResultDataAccessException e) {
+					//somebody deleted it between calls. try to insert again (go to beginning of while loop)
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean replace(String key, String oldValue, String newValue) {
```

Comment:
Since this single query does not need transnational behavior? All other queries are transnational since we need to lock a row for editing and atomic operations. Here it is only replace and we don't care if it is not successful (affectedRows== 0).